crypto/slh_dsa/slh_dsa.c

```diff
@@ -0,0 +1,359 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+#include <assert.h>
+#include <stddef.h>
+#include <string.h>
+#include <openssl/err.h>
+#include <openssl/proverr.h>
+#include "slh_dsa_local.h"
+#include "slh_dsa_key.h"
+
+#define SLH_MAX_M 49 /* See slh_params.c */
+/* The size of md is (21..40 bytes) - since a is in bits round up to nearest byte */
+#define MD_LEN(params) (((params)->k * (params)->a + 7) >> 3)
+
+static int get_tree_ids(PACKET *pkt, const SLH_DSA_PARAMS *params,
+                        uint64_t *tree_id, uint32_t *leaf_id);
+
+/**
+ * @brief SLH-DSA Signature generation
+ * See FIPS 205 Section 9.2 Algorithm 19
+ *
+ * A signature consists of
+ *   r[n] random bytes
+ *   [k]*[1+a][n] FORS signature bytes
+ *   [h + d*len][n] Hyper tree signature bytes
+ *
+ * @param ctx Contains SLH_DSA algorithm functions and constants, and the
+ *            private SLH_DSA key to use for signing.
+ * @param msg The message to sign. This may be encoded beforehand.
+ * @param msg_len The size of |msg|
+ * @param sig The returned signature
+ * @param sig_len The size of the returned |sig|
+ * @param sig_size The maximum size of |sig|
+ * @param opt_rand An optional random value to use of size |n|. It can be NULL.
+ * @returns 1 if the signature generation succeeded or 0 otherwise.
+ */
+static int slh_sign_internal(SLH_DSA_HASH_CTX *hctx,
+                             const uint8_t *msg, size_t msg_len,
+                             uint8_t *sig, size_t *sig_len, size_t sig_size,
+                             const uint8_t *opt_rand)
+{
+    int ret = 0;
+    const SLH_DSA_KEY *priv = hctx->key;
+    const SLH_DSA_PARAMS *params = priv->params;
+    size_t sig_len_expected = params->sig_len;
+    uint8_t m_digest[SLH_MAX_M];
+    const uint8_t *md; /* The first md_len bytes of m_digest */
+    size_t md_len = MD_LEN(params); /* The size of the digest |md| */
+    /* Points to |m_digest| buffer, it is also reused to point to |sig_fors| */
+    PACKET r_packet, *rpkt = &r_packet;
+    uint8_t *r, *sig_fors; /* Pointers into buffer inside |wpkt| */
+    WPACKET w_packet, *wpkt = &w_packet; /* Points to output |sig| buffer */
+    const uint8_t *pk_seed, *sk_seed; /* pointers to elements within |priv| */
+    uint8_t pk_fors[SLH_MAX_N];
+    uint64_t tree_id;
+    uint32_t leaf_id;
+
+    SLH_ADRS_DECLARE(adrs);
+    SLH_HASH_FUNC_DECLARE(priv, hashf);
+    SLH_ADRS_FUNC_DECLARE(priv, adrsf);
+
+    if (sig_len != NULL)
+        *sig_len = sig_len_expected;
+
+    if (sig == NULL)
+        return (sig_len != NULL);
```

Comment:
Is this consistent with our approach in the other sig implementations for not raising errors on these checks here?

Reply:
Others dont even check sig_len so I will remove the check and rearrange..